docs/architecture/adr-028-public-key-addresses.md

```diff
@@ -3,49 +3,90 @@
 
-In discussions in [\#5694](https://github.com/cosmos/cosmos-sdk/issues/5694), we agreed to go with an
-approach similar to this where essentially we take the first 20 bytes of the `sha256` hash of
-the key type concatenated with the key bytes, summarized as `Sha256(KeyTypePrefix || Keybytes)[:20]`.
+This lead to the first proposal (which we proved to be not good enough):
+we concatenate a key type with a public key, hash it and take the first 20 bytes of that hash, summarized as `sha256(keyTypePrefix || keybytes)[:20]`.
+
+
+### Review and Discussions
+
+In [\#5694](https://github.com/cosmos/cosmos-sdk/issues/5694) we discussed various solutions.
+We agreed that 20 bytes it's not future proof, and extending the address length is the only way to allow addresses of different types, various signature types, etc.
+This disqualifies the initial proposal.
+
+In the issue we discussed various modifications:
++ Choice of the hash function.
++ Move the prefix out of the hash function: `keyTypePrefix + sha256(keybytes)[:20]` [post-hash-prefix-proposal].
++ Use double hashing: `sha256(keyTypePrefix + sha256(keybytes)[:20])`.
++ Increase to keybytes hash slice from 20 byte to 32 or 40 bytes. We concluded that 32 bytes, produced by a good hash functions is future secure.
+
+### Requirements
+
++ Support currently used tools - we don't want to break an ecosystem, or add a long adaptation period. Ref: https://github.com/cosmos/cosmos-sdk/issues/8041
++ Try to keep the address length small - addresses are widely used in state, both as part of a key and object value.
+
+
+### Scope
+
+This ADR only defines a process for the generation of address bytes. For end-user interactions with addresses (through the API, or CLI, etc.), we still use bech32 to format these addresses as strings. This ADR doesn't change that.
+Using bech32 for string encoding gives us support for checsum error codes and handling of user typos.
+
 
 ## Decision
 
+We define the following account types, for which we define the address function:
+
+1. simple accounts: represented by a regular public key (ie: secp256k1, sr25519)
+2. naive multisig: accounts composed by other addressable objects (ie: naive multisig)
+3. composed accounts with a native address key (ie: bls, group module accounts)
+4. module accounts: basically any accounts which cannot sign transactions and which are managed internally by modules
+
+
 ### Legacy Public Key Addresses Don't Change
 
-`secp256k1` and multisig public keys are currently in use in existing Cosmos SDK zones. They use the following
-address formats:
+Currently (Jan 2021), the only officially supported SDK user accounts are `secp256k1` basic accounts and legacy amino multisig.
```

Comment:
we have ed25519 also, and the address is `sha256(pubkey)[:20]`

Reply:
the reason we want to push this ASAP is to not include ed25519 in legacy.

Reply:
NOTE, ed25519 are not supported by antehandler.